### PR TITLE
[server, protocol] Further fix and improve API tracing

### PR DESCRIPTION
**Pull request by geropl**
## Description

This started with a fix for [this flaw](https://github.com/gitpod-io/gitpod/pull/6995#discussion_r760498848), but grew a bit bigger. It contains:
 - fix the original bug (API parameters being mangled) ([commit](https://github.com/gitpod-io/gitpod/pull/7032/commits/a486971274f61ca81975563c05a70d48e29d2629))
 - add `redactJsonRPCParameters` to opt-out of tracing of certain API arguments (because it contains personal indentifiable information, for instance) ([commit](https://github.com/gitpod-io/gitpod/pull/7032/commits/fadbf38e9da9813bfe1bd71427415af563071074))
 - `TraceContext.setOWI` as analogon to the method we have in Go-components. It adds "owner-workspace-instance" (and session) ids to logs and traces, to make them discoverable across all components/debugging tools. ([commit](https://github.com/gitpod-io/gitpod/pull/7032/commits/33c50c3ac87eb37d8d81ef945af9764fab044cd6))
 - apply both `setOWI` and `redactJsonRPCParameters` to GitpodServerImpls ([commit](https://github.com/gitpod-io/gitpod/pull/7032/commits/6241cfb6fc668146e51e701fdd8b85dd5151b728))
 - add `sessionId` to be added to every call ([commit](https://github.com/gitpod-io/gitpod/pull/7032/commits/f9a219a20281cbeeeba6bf84516571a31bb651a7))

## Related Issue(s)
Context: #6995

## How to test
Not sure how to test this in devstaging...?

## Release Notes

```release-note
NONE
```

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **geropl** after the PR has been reviewed.

Associated issue: *#6995*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by geropl** [comment]
/werft run

There was a race in the status reporting from werft :roll_eyes:

File: components/gitpod-protocol/src/util/tracing.ts
Comment (geropl) [inline components/gitpod-protocol/src/util/tracing.ts:161]: This is sth we definitely need to align with the Go components (not namespaced there!). I'm definitely in favor of having `owi` namespaced somehow because:
 - it eases browsing of the fields in honeycomb
 - more unlikely to be completely overriden by a random `setTag`

No hard feelings about the name, though.

**Comment by jankeromnes** [comment]
Not feeling super confident reviewing these changes, but I'll give them a look! 👀

Many thanks for writing up helpful details in the PR description. 💯

**Comment by geropl** [comment]
@jankeromnes It might help reviewing it commit by commit! This also highlight the interesting stuff, and separates it from the 4th one, which is very noisy. :roll_eyes:

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [inline components/server/src/workspace/gitpod-server-impl.ts:391]: No `traceWI(ctx, { workspaceId });`?

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [inline components/server/src/workspace/gitpod-server-impl.ts:558]: Oh, isn't this RPC parameter already automatically traced?

**Review by jankeromnes — COMMENTED** [review]
Okay, let's call this first round done!

Many thanks for improving the tracing in server. I mostly have a general question about opt-in vs opt-out, and the implications on accidentally logging PII / other sensitive data.

File: components/gitpod-protocol/src/util/tracing.ts
Comment (geropl) [reply components/gitpod-protocol/src/util/tracing.ts:113]: I can share some of my thoughts, but hope @mads-hartmann has some thoughts and experience to share as well! :smile: 

Regarding PII in general, especially the question "what if this harmless field contains PII" (example: `but what if layoutData accidentally contains PII?`)
 - PII is about the fields that are semantically known to contain PII, like `name`, `address`, `githubUserId`, which can be used (in combination with publicly available sources, like GitHub) to identify you in person
 - it's _not_ about user (mis-)using other fields to put their name, address, etc. Imagine a repo named `<myfullname>_<myaddress>`: we cannot be held accountable to censor this; that's impossible
 - that being said, there might be cases where we'd like to err on the side of caution. But that again is independent from "opt-in vs opt-out"; as our goal is to trace as much as possible we have to decide in either case. `layoutData` might be such a case; I personally have no clue what's in there. :shrug: 

Regarding the general choice "opt-in vs opt-out":
 - I personally have no hard feelings here. I started this PR with opt-in because of [this comment](https://github.com/gitpod-io/gitpod/pull/6995#pullrequestreview-820266817) from @mads-hartmann how has some experience with tracing. Having a single line at the beginning of each method is also fine with me. Would safe us from the strange "indentify by index" problem (cmp. `redactJsonRPCParameters`)
 - to re-iterate:
   - we have to be careful with PII either way, every time we change the API (nested shapes)
   - it's desirable to trace as much as possible (as long as it's not "huge data"). This is different to before, where we'd be cautious with what we trace because how badly our self-managed jaeger handled the (small) amount of data. So comparing to "how it was before" is not a good guideline: traces where near-unusable, and not present for most API calls in the first place. :slightly_smiling_face: 

> UserEnvVarValue in setEnvVar should probably always be censored

> GuessGitTokenScopesParams contains a GitToken

Both true :100:, nice findings. :pray: Will fix.

> Is the content part of UpdateUserStorageResourceOptions always safe to trace in updateUserStorageResource?

Content is already excluded I think. Will verify.

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (geropl) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:1061]: See [comment above](https://github.com/gitpod-io/gitpod/pull/7032#discussion_r761889048), especially this part:

> Having a single line at the beginning of each method is also fine with me. Would safe us from the strange "indentify by index" problem (cmp. redactJsonRPCParameters). AND would give us "named parameters", so rpc.jsonrpc.parameters.workspaceId instead of rpc.jsonrpc.parameters.0. Especially the last two make me reconsider... :thinking:

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (geropl) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:620]: Where's my answer gone??? Ok, here's the short version:
 - `this` inside `GitpodServerImpl` still points to the "regular" object
 - only API calls (public methods called from the websocket) have their parameters traced because they don't call `GitpodServerImpl` directly but through [this code](https://github.com/gitpod-io/gitpod/blob/952bc6dd24c4e682f4c3c3fd308cded64b93984d/components/server/src/websocket/websocket-connection-manager.ts#L285-L347)

File: components/gitpod-protocol/src/util/tracing.ts
Comment (mads-hartmann) [reply components/gitpod-protocol/src/util/tracing.ts:161]: So the way I'd usually go would multiple levels of namespacing. E.g. `workspace.owner.instance` that way you have something quite discoverable in your tool - you start writing `workspace.` and you'll get all attributes related to workspace.

Comment (geropl) [reply components/gitpod-protocol/src/util/tracing.ts:161]: > So the way I'd usually go would multiple levels of namespacing

I understand the general id. But our core data structure is very flat (there is not much beyond user -> workspace -> workspaceInstance). And we already use this shape (with namespace `context`) in logging, for instance, for years. Also, all Go components use this for tracing (albeit not namespace - yet).

Happy to duplicate this field, or have additional convetions, but this one just makes so much sense. :-)

File: components/gitpod-protocol/src/util/tracing.ts
Comment (mads-hartmann) [reply components/gitpod-protocol/src/util/tracing.ts:113]: I can extend a bit about the motivations behind adding all parameters automatically, if you don't mind a little detour.

There isn't a single agreed-upon definition of observability. Back in [august 2019](https://blog.mads-hartmann.com/sre/2019/08/04/journey-into-observability-reading-material.html) I had the following take (sorry for quoting myself 😂) based on what I had read and my experience at the time.

> Observability is all about being able to ask questions of your system and get answers based on the existing telemetry it produces; if you have to re-configure or modify the service to get answers to your questions you haven’t achieved observability yet

I think it's still a good guiding definition to consider when you instrument your services. You want to strive towards having all the information you need available to you when you're debugging, and sometimes you can't know what that information is ahead of time. With metrics this is impossible as each label you add will introduce more time series which explodes based on the number of labels and their cardinality. With traces that's not a concern, so you're free to add as many attributes as you like. So I'd always default to add things automatically assuming that you can namespace it correctly.

So in that sense, having to specify exactly which parameter to add to the traces means you have to think about what information you might need during debugging ahead of time, and you will inevitably overlook something important.

The PII concern is important though. Honeycomb has a retention of 60 so the data is gone after that, but it would still have reached their servers of course. 

It's a shame that JSON RPC arguments are positional, otherwise we might have been able to add some processing in the OpenTelemetry Collector to redact any parameter name that matched "email" or "address" etc.

I'd lean towards the opt-out approach as the vast majority of your parameters (I'm assuming) isn't PII. And as you already have to be careful about PII I don't think it's too much additional load to think about redacting it from telemetry.

But this is up to you. Having a single line in each method also allows you to name the parameters as @geropl highlighted, so that has benefits too. Either way this is a huge improvement and I'm happy to see it.

Comment (mads-hartmann) [reply components/gitpod-protocol/src/util/tracing.ts:113]: After having read through the rest of the comments here I agree with @jankeromnes that redacting based on the position of the argument is fragile and I change my recommendation to be that you add a line where you opt-in to adding the arguments, which also gives you the opportunity to give them names, as you suggested @geropl ☺️

For the future, perhaps we can find a way to use named arguments with JSON RPC so we can get the best of both worlds.

Comment (geropl) [reply components/gitpod-protocol/src/util/tracing.ts:113]: Thx @mads-hartmann ! :+1: make sense.


> For the future, perhaps we can find a way to use named arguments with JSON RPC so we can get the best of both worlds.

Hopefully we can replace it with something else; maybe gRPC. :smiley:

File: components/gitpod-protocol/src/util/tracing.ts
Comment (mads-hartmann) [reply components/gitpod-protocol/src/util/tracing.ts:161]: For now, go with whatever is most consistent with what you have already. We'll then together across all teams work towards standardising on namespaces, but right now only worry about namespaces when you're introducing new attributes. I'm writing up the initial document now so we'll have something to iterate on  ☺️

**Comment by geropl** [comment]
@jankeromnes Ready for a second round of review. Now with opt-in params tracing with names: `traceAPIParams(ctx, { workspaceId, duration });` :slightly_smiling_face: IMO a fair tradeoff between conciseness, safety and better tracing.

**Comment by jankeromnes** [comment]
> Ready for a second round of review.

Awesome! Many thanks for reworking, will have a look soon.

Also, in the future, please re-request review in the top right of the PR. This helps me easily differentiate PRs between "work in progress" and "blocked until reviewed". 😇

> Now with opt-in params tracing with names: `traceAPIParams(ctx, { workspaceId, duration });` :slightly_smiling_face: IMO a fair tradeoff between conciseness, safety and better tracing.

Wow, thank you a lot for switching to opt-in! 🙏 I agree that this trade-off seems like our best option available. 👍

**Review by mads-hartmann — APPROVED** [review]
I can't know much about server, so I focused on the instrumentation implementation and the naming conventions used. This all looks great to me!

And thanks for adding tests! 🙏

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7032#pullrequestreview-823771968" title="Approved">mads-hartmann</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **geropl** after the PR has been reviewed.

Associated issue: *#6995*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Review by jankeromnes — COMMENTED** [review]
Alright, here comes round 2! 😁

It's actually mostly the final round, I'd say we're at "LGTM minus the question about whether or not to log `.gitpod.yml`".

Sorry for the many in-line comments 🙈 -- they're mostly praise or questions, but I think there are a few items that merit at least some clarification before we merge. 🚀

File: components/gitpod-protocol/src/util/tracing.ts
Comment (geropl) [reply components/gitpod-protocol/src/util/tracing.ts:68]: Hint for the future :+1:

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (geropl) [reply components/server/src/workspace/gitpod-server-impl.ts:59]: Thought about this as well. But IMO there is a semantic difference between a `workspaceId` parameter being traced and an being reported as `OWI` (owner-workspace-instance) - at least in my head. :stuck_out_tongue: 
 - OWI is about the user-workspace-instance at hand, as specific as possible, but _never_ mismatching non-owner `userId` with `workspaceId`
 - parameters is just parameters, it might related to any user/workspace

That being said, there is no documentation about this anywhere, yet. Mads started to write one, and I hope to have that discussion there.

I'd like to keep it as-is for now, clarify the tracing documentation in notion, and than maybe align later.

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (geropl) [reply components/server/src/workspace/gitpod-server-impl.ts:60]: `censor` is nice! :+1:

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (geropl) [reply components/server/src/workspace/gitpod-server-impl.ts:1639]: People _are_ putting keys to external services into `.gitpod.yml` files... ;-)

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (geropl) [reply components/server/src/workspace/gitpod-server-impl.ts:1683]: No. There is a handful of methods I see potential for cleanup for. Different PR :slightly_smiling_face:

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (geropl) [reply components/server/src/workspace/gitpod-server-impl.ts:1559]: Separate PR :slightly_smiling_face:

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (geropl) [reply components/server/src/workspace/gitpod-server-impl.ts:2084]: metrics !== tracing

Histograms for API call duration is something we should do with metrics - which we still don't have I think :roll_eyes:

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (geropl) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:403]: Not sure how to treat search fields. Everybody can put everything into it. Does it even work with email? :thinking:

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (geropl) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:717]: No, excellent point! Will remove.

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (geropl) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:1408]: IMO this is still (ab-)used... from admin panel, maybe?

File: components/gitpod-protocol/src/util/tracing.ts
Comment (geropl) [reply components/gitpod-protocol/src/util/tracing.ts:32]: That's personal taste I think. :slightly_smiling_face:

**Comment by roboquat** [comment]
New changes are detected. LGTM label has been removed.

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (geropl) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:403]: It actuall does :laugh:

**Comment by geropl** [comment]
@jankeromnes Went through all comments again, and think answered them all.

**Comment by jankeromnes** [comment]
Many thanks! Taking another look 👀

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [reply components/server/src/workspace/gitpod-server-impl.ts:59]: > I'd like to keep it as-is for now, clarify the tracing documentation in notion, and than maybe align later.

Thanks for explaining your thought process! Works for me 👍 🚀

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [reply components/server/src/workspace/gitpod-server-impl.ts]: Okay, fair enough. 👍

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [reply components/server/src/workspace/gitpod-server-impl.ts:2091]: I think tracing is still relevant for performance (e.g. to investigate "why does it take so long to start a workspace?" -- here API call durations only give you part of the story I think, and investigating the sequence of events on the hot path with timings is helpful).

Anyway, let's not worry about hypothetical problems, we have enough real ones! 😂

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [reply components/server/ee/src/workspace/gitpod-server-impl.ts]: Yes, I think it was designed that way (to find users by email address)

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:1405]: Doesn't seem to be called from anywhere: https://github.com/gitpod-io/gitpod/search?q=sendFeedback

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [inline components/server/src/workspace/gitpod-server-impl.ts:1818]: **Nit**: Cool trick censoring `config` then partially adding it back!

However, I think you're changing the semantics here (e.g. if `config` was `undefined`, in tracing it suddenly becomes defined as an object like `config = { vscode: undefined }`.

Maybe safer like this:

```suggestion
                ...censor(params, "config"),
                ...(!!params.config ? : {
                    config: {
                        vscode: params.config.vscode,
                    },
                } : {}),
```

**Review by jankeromnes — APPROVED** [review]
Many thanks! 🙏

Approved with one nit: The semantic change on the traced `ResolvePluginsParams` (please see in-line).

Summarizing follow-ups for later:
- [x] Maybe `getPortAuthenticationToken` needs a `traceWI()`? https://github.com/gitpod-io/gitpod/pull/7032#discussion_r761844358
- [ ] Remove unused `userId` argument in `deleteTeam` https://github.com/gitpod-io/gitpod/pull/7032#discussion_r762876384
- [ ] Audit `name` in call sites of `getContentBlobUploadUrl`, or delete this method if it's no longer used https://github.com/gitpod-io/gitpod/pull/7032#discussion_r762883065
- [ ] Delete ancient unused `sendFeedback` server method https://github.com/gitpod-io/gitpod/pull/7032#discussion_r763005329

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 2ad256e00154102e9e8610a7ca368031c25a1abe</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7032#pullrequestreview-824254136" title="Approved">jankeromnes</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/7032#pullrequestreview-823771968" title="Approved">mads-hartmann</a>*

Associated issue: *#6995*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [jankeromnes]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [jankeromnes]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [jankeromnes]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
New changes are detected. LGTM label has been removed.

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (geropl) [reply components/server/src/workspace/gitpod-server-impl.ts]: Yep. Went with conciseness over preciseness here... but now that you found it: here you go! :wink:

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [inline components/server/src/workspace/gitpod-server-impl.ts:1817]: Ah, I think checking `param.config` was more correct than checking `params.config.vscode`, but oh well.

We've done enough hair-splitting! 😇

**Review by jankeromnes — APPROVED** [review]
Still looks good to me! 🚀

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 47f8a9cf9274d3a80451ae5b52217dd17a2edf0f</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7032#pullrequestreview-824942729" title="Approved">jankeromnes</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/7032#pullrequestreview-823771968" title="Approved">mads-hartmann</a>*

Associated issue: *#6995*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [jankeromnes]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [jankeromnes]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [jankeromnes]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
